src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py

```diff
@@ -501,9 +503,114 @@
+    method_name: str
+
+
+# Rule defining which args should trigger auto-switching.
+@dataclass
+class UnsupportedArgsRule:
+    """
+    Rule for defining argument combinations that trigger auto-switching to pandas.
+
+    Attributes:
+        unsupported_conditions: List of conditions that can be either:
+            - tuple[Callable, str]: (condition_function, reason) for complex conditions
+            - tuple[str, Any]: (argument_name, unsupported_value) for simple value checks
+    """
+
+    unsupported_conditions: List[Union[Tuple[Callable, str], Tuple[str, Any]]] = field(
+        default_factory=list
+    )
+
+
+# Set of MethodKey objects for methods that are wholly unimplemented by
 # Snowpark pandas. This list is populated by the register_*_not_implemented decorators.
-HYBRID_SWITCH_FOR_UNIMPLEMENTED_METHODS: Set[Tuple[str, str]] = set()
+HYBRID_SWITCH_FOR_UNIMPLEMENTED_METHODS: Set[MethodKey] = set()
+
+# Global registry for args-based switching rules
+HYBRID_SWITCH_FOR_UNSUPPORTED_ARGS: dict[MethodKey, UnsupportedArgsRule] = {}
+
+
+def register_query_compiler_method_not_implemented(
+    api_cls_name: Optional[str],
+    method_name: str,
+    unsupported_args: Optional["UnsupportedArgsRule"] = None,
+) -> Callable[[Callable[..., Any]], Callable[..., Any]]:
+    """
+    Decorator for SnowflakeQueryCompiler methods with args-based auto-switching.
+
+    Registers pre-op switching and raises a NotImplementedError if any unsupported-parameter predicate evaluates True.
+
+    Args:
+        api_cls_name: Frontend class name (e.g., "BasePandasDataset", "Series", "DataFrame", "None").
+        method_name: Method name to register.
+        unsupported_args: UnsupportedArgsRule for args-based auto-switching.
+                            If None, method is treated as completely unimplemented.
+    """
+    reg_key = MethodKey(api_cls_name, method_name)
+
+    # register the method in the hybrid switch for unsupported args
+    if unsupported_args is None:
+        HYBRID_SWITCH_FOR_UNIMPLEMENTED_METHODS.add(reg_key)
+    else:
+        HYBRID_SWITCH_FOR_UNSUPPORTED_ARGS[reg_key] = unsupported_args
+
+    from modin.core.storage_formats.pandas.query_compiler_caster import (
+        register_function_for_pre_op_switch,
+    )
+
+    register_function_for_pre_op_switch(
+        class_name=api_cls_name, backend="Snowflake", method=method_name
+    )
+
+    def decorator(query_compiler_method: Callable[..., Any]) -> Callable[..., Any]:
+        @functools.wraps(query_compiler_method)
+        def wrapper(self: "SnowflakeQueryCompiler", *args: Any, **kwargs: Any) -> Any:
+            bound_arguments = inspect.signature(query_compiler_method).bind(
+                self, *args, **kwargs
+            )
+            bound_arguments.apply_defaults()
+            # Extract parameters excluding 'self'
+            arguments = MappingProxyType(
+                {k: v for k, v in bound_arguments.arguments.items() if k != "self"}
+            )
```

Comment:
```suggestion
            # Extract parameters excluding 'self'
            arguments = bound_arguments.arguments
            del arguments["self"]
```
nit: a little easier to read (assuming `bound_arguments` isn't needed anywhere later)